### ISSUE TITLE
How to apply MDMLP to 1D[B,C,H]?

### ISSUE THREAD

**Issue by kuaileyuandi** [Amoza-Theodore/MDMLP#1]
Hello, Thank you very much for conceiving the MDMLP model. Not only the amount of parameters is very small, but also a high accuracy rate is obtained, which is very innovative. Now MDMLP is at two-dimensional level [B,C,H,W], but when the data is one-dimensional data [B,C,H], how to use MDMLP model? I would like to ask a question: Can you provide an MDMLP with a 1D MLP level for reference?

**Comment by PhD-TianLv** [comment]
Hello, nice to see my work being used by others, here's a reference.

```python
import torch
import torch.nn as nn
from einops import rearrange
from einops.layers.torch import Rearrange, Reduce
from collections import OrderedDict


class OverlapPatchEmbed(nn.Module):
    def __init__(self, patch_size, overlap=0):
        super(OverlapPatchEmbed, self).__init__()
        self.patch_size, self.overlap = patch_size, overlap

    def forward(self, x):
        batch_size, in_channs, img_size = x.shape
        patch_size, overlap = self.patch_size, self.overlap

        stride = patch_size - overlap
        patch_num = (img_size - patch_size) // stride + 1

        patches = torch.zeros([batch_size, patch_num, in_channs, patch_size], device=x.device)
        for i in range(patch_num):
            # x: [B, C, H]
            patch = x[:, :, i * stride:i * stride + patch_size]
            patches[:, i, :, :] = patch

        return patches  # [B, H', C, D]


class MLPLayer(nn.Module):
    def __init__(self, base_dim, dim, factor, forw_perm, **axes_lengths):
        super(MLPLayer, self).__init__()
        back_perm = forw_perm.split("->")[1] + " -> " + forw_perm.split("->")[0]

        self.norm = nn.LayerNorm(base_dim)

        self.forw_rearr = Rearrange(forw_perm, **axes_lengths)
        self.dense1 = nn.Linear(dim, int(dim * factor))
        self.dense2 = nn.Linear(int(dim * factor), dim)
        self.gelu = nn.GELU()
        self.back_rearr = Rearrange(back_perm, **axes_lengths)

    def forward(self, x):
        y = x

        y = self.norm(y)

        y = self.forw_rearr(y)
        y = self.dense1(y)
        y = self.gelu(y)
        y = self.dense2(y)
        y = self.gelu(y)
        y = self.back_rearr(y)

        y = y + x
        return y


class MDBlock(nn.Module):
    def __init__(self, patch_num, channs, base_dim, factor):
        super(MDBlock, self).__init__()
        h = w = patch_num
        self.hlayer = MLPLayer(base_dim, patch_num, factor, "b h c d -> b c d h", h=h)
        self.clayer = MLPLayer(base_dim, channs, factor, 'b h c d -> b h d c', h=h)
        self.dlayer = MLPLayer(base_dim, base_dim, factor, 'b h c d -> b h c d', h=h)

    def forward(self, x):
        y = self.hlayer(x)
        y = self.clayer(y)
        y = self.dlayer(y)
        return y


class MDMLP(nn.Module):
    def __init__(self,
                 img_size,
                 in_chans,
                 num_classes,
                 base_dim,
                 depth,
                 patch_size,
                 overlap,
                 factor=4,
                 **kwargs):
        super(MDMLP, self).__init__()
        self.num_classes = num_classes
        patch_num = (img_size - patch_size) // (patch_size - overlap) + 1
        channs = in_chans

        self.stem = nn.Sequential(OrderedDict([
            ('OverlapPatchEmbed', OverlapPatchEmbed(patch_size, overlap)),
            ('Linear', nn.Linear(patch_size, base_dim)),
        ]))

        self.blocks = nn.Sequential(OrderedDict([
            *[(f'MDLayer{d}', MDBlock(patch_num, channs, base_dim, factor))
              for d in range(depth)]
        ]))

        self.head = nn.Sequential(OrderedDict([
            ('Reduce', Reduce('b p c d -> b d', 'mean')),
            ('Linear', nn.Linear(base_dim, num_classes)),
        ]))

    def forward(self, x):
        x = self.stem(x)
        x = self.blocks(x)
        x = self.head(x)

        return x


if __name__ == '__main__':
    x = torch.randn([2, 3, 128])  # [B, C, H]
    model = MDMLP(
        img_size=128,
        in_chans=3,
        num_classes=10,
        base_dim=64,
        depth=4,
        patch_size=16,
        overlap=0,
    )
    y = model(x)
    pass
```
